### PR TITLE
8301036: RISC-V: Factor out functions baseOffset & baseOffset32 from MacroAssembler

### PR DESCRIPTION
**Pull request by RealFYang**
On RISC-V, functions baseOffset & baseOffset32 from MacroAssembler receive an Address in base_plus_offset mode. These two functions check the range of the offset, add it and base and put the result in a destination register. This duplicates function MacroAssembler::la in functionality. We could refactor this part moving the logic of Address legitimization into MacroAssembler::la and use this function instead.

Testing:
  - [x] Bootcycle (release & fastdebug)
  - [x] Tier1-4 (release)
  - [x] Benchmarks: SPECjbb2015 (release)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301036](https://bugs.openjdk.org/browse/JDK-8301036): RISC-V: Factor out functions baseOffset & baseOffset32 from MacroAssembler

### Reviewers
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Committer) ⚠️ Review applies to [c617b7f8](https://git.openjdk.org/jdk/pull/12177/files/c617b7f88e0dec3b7f6061f43f4cd88364e92335)
 * [Feilong Jiang](https://openjdk.org/census#fjiang) (@feilongjiang - Author)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12177/head:pull/12177` \
`$ git checkout pull/12177`

Update a local copy of the PR: \
`$ git checkout pull/12177` \
`$ git pull https://git.openjdk.org/jdk pull/12177/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12177`

View PR using the GUI difftool: \
`$ git pr show -t 12177`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12177.diff">https://git.openjdk.org/jdk/pull/12177.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 1]
```diff
@@ -2411,11 +2385,14 @@ void MacroAssembler::membar(uint32_t order_constraint) {
...
 // is up to you to ensure that the shift provided matches the size
 // of your data.
-Address MacroAssembler::form_address(Register Rd, Register base, long byte_offset) {
+Address MacroAssembler::form_address(Register Rd, Register base, int64_t byte_offset) {
   if (is_offset_in_range(byte_offset, 12)) { // 12: imm in range 2^12
     return Address(base, byte_offset);
   }
 
+  assert(Rd != noreg, "expecting a register");
+  guarantee(Rd != base, "should be different registers");
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2394]: Why not `assert`?

Comment (RealFYang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Fixed. These two lines come from the original MacroAssembler::baseOffset32. I think assert_different_registers will also do here.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 2]
```diff
@@ -2411,11 +2385,14 @@ void MacroAssembler::membar(uint32_t order_constraint) {
...
 // is up to you to ensure that the shift provided matches the size
 // of your data.
-Address MacroAssembler::form_address(Register Rd, Register base, long byte_offset) {
+Address MacroAssembler::form_address(Register Rd, Register base, int64_t byte_offset) {
   if (is_offset_in_range(byte_offset, 12)) { // 12: imm in range 2^12
     return Address(base, byte_offset);
   }
 
+  assert(Rd != noreg, "expecting a register");
+  assert_different_registers(Rd, base);
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2394]: You can also do `assert_different_registers(Rd, base, noreg);` to cut down on the checks, except if `base` can be a `noreg`.

**Review by luhenry — APPROVED** [review]

**Review by feilongjiang — APPROVED** [review]

**Review by shipilev — APPROVED** [review]
Looks fine.

**Comment by RealFYang** [comment]
Thanks all for the review. I don't think the windows-x64 pre-submit failure is relevant here.
/integrate

**Pull request closed without merging**